neo/SmartContract/Native/Tokens/NeoToken.cs

```diff
@@ -0,0 +1,422 @@
+                case "getValidators":
+                    result = GetValidators(engine).Select(p => (StackItem)p.ToArray()).ToArray();
+                    break;
+                default:
+                    return false;
+            }
+            engine.CurrentContext.EvaluationStack.Push(result);
+            return true;
+        }
+
+        private static StorageKey CreateStorageKey(UInt160 script_hash, byte prefix, byte[] key = null)
+        {
+            StorageKey storageKey = new StorageKey
+            {
+                ScriptHash = script_hash,
+                Key = new byte[sizeof(byte) + (key?.Length ?? 0)]
+            };
+            storageKey.Key[0] = prefix;
+            if (key != null)
+                Buffer.BlockCopy(key, 0, storageKey.Key, 1, key.Length);
+            return storageKey;
+        }
+
+        private static BigInteger BalanceOf(ApplicationEngine engine, byte[] account)
+        {
+            if (account.Length != 20) throw new ArgumentException();
+            StorageItem storage = engine.Service.Snapshot.Storages.TryGet(CreateStorageKey(ScriptHash, Prefix_Account, account));
+            if (storage is null) return BigInteger.Zero;
+            Struct state = (Struct)storage.Value.DeserializeStackItem(engine.MaxArraySize);
+            return state[0].GetBigInteger();
+        }
+
+        private static bool Transfer(ApplicationEngine engine, byte[] from, byte[] to, BigInteger amount)
+        {
+            if (engine.Service.Trigger != TriggerType.Application) throw new InvalidOperationException();
+            UInt160 hash_from = new UInt160(from);
+            UInt160 hash_to = new UInt160(to);
+            if (amount.Sign < 0) throw new ArgumentOutOfRangeException(nameof(amount));
+            if (!hash_from.Equals(new UInt160(engine.CurrentContext.CallingScriptHash)) && !engine.Service.CheckWitness(engine, new UInt160(from)))
+                return false;
+            ContractState contract_to = engine.Service.Snapshot.Contracts.TryGet(hash_to);
+            if (contract_to?.Payable == false) return false;
+            StorageKey key_from = CreateStorageKey(ScriptHash, Prefix_Account, from);
+            StorageItem storage_from = engine.Service.Snapshot.Storages.TryGet(key_from);
+            if (amount.IsZero)
+            {
+                if (storage_from != null)
+                {
+                    AccountState state_from = AccountState.FromByteArray(storage_from.Value);
+                    DistributeGas(engine, from, state_from);
+                    storage_from = engine.Service.Snapshot.Storages.GetAndChange(key_from);
+                    storage_from.Value = state_from.ToByteArray();
+                }
+            }
+            else
+            {
+                if (storage_from is null) return false;
+                AccountState state_from = AccountState.FromByteArray(storage_from.Value);
+                if (state_from.Balance < amount) return false;
```

Comment:
In which case this if will happen?
Maybe after tests this can be removed, right?

Reply:
Insufficient balance

Reply:
True, there isn't another place for to verify this before. 
Why to proceed normally and return `true` when amount is zero (line 119 and then reaching 192)?

Reply:
https://github.com/neo-project/proposals/blob/master/nep-5.mediawiki#transfer

> The parameter `amount` MUST be greater than or equal to `0`. If not, this method SHOULD throw an exception.

> If the method succeeds, it MUST fire the `transfer` event, and MUST return `true`, even if the `amount` is `0`, or `from` and `to` are the same address.

Reply:
`The parameter amount MUST be greater than or equal to 0.`
Ok, this is strange, but if it is like this.